### PR TITLE
adapter: Implement MVP for real time recency

### PR DESCRIPTION
**Pull request by jkosh44**
This commit implements an MVP for real time recency on the adapter side, gated behind unsafe mode.

Currently, the Coordinator waits for a recent timestamp in the main coord loop, which can potentially block the Coordinator for an unbounded amount of time. Before unsafe mode is removed, this waiting will have to be removed from the main Coordinator loop.

Works towards resolving #11531

### Motivation
This PR adds a known-desirable feature.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

File: src/controller/src/lib.rs [hunk 1]
```diff
@@ -217,7 +217,8 @@ where
     /// `source_ids` at the time of the function call.
     #[allow(unused)]
     #[allow(clippy::unused_async)]
-    pub async fn recent_timestamp(&self, source_ids: &[GlobalId]) -> T {
+    pub async fn recent_timestamp(&self, source_ids: impl Iterator<Item = GlobalId>) -> T {
```
Comment (jkosh44) [inline src/controller/src/lib.rs:220]: @petrosagg Are you ok with this API change?

Comment (petrosagg) [reply src/controller/src/lib.rs:220]: We don't yet have an implementation so.. tentative yes. Once we start wiring up the storage side we can see if we need to bring back the slice. Worst case we collect into a vec. Feel free to merge it for now

File: src/adapter/src/coord/sequencer.rs [hunk 2]
```diff
@@ -2387,6 +2399,31 @@ impl<S: Append + 'static> Coordinator<S> {
         })
     }
 
+    async fn recent_timestamp(
```
Comment (mjibson) [inline src/adapter/src/coord/sequencer.rs:2402]: Can this get at least a vague doc comment describing the intent of the function and its asyncness?

File: src/adapter/src/session.rs [hunk 3]
```diff
@@ -428,6 +428,20 @@ impl<T: TimestampManipulation> Session<T> {
...
+        match self.transaction.inner() {
+            Some(Transaction {
+                pcx: _,
+                ops: TransactionOps::Peeks(TimestampContext::TimelineTimestamp(_, _)),
+                write_lock_guard: _,
+                access: _,
+                id: _,
+            }) => true,
+            _ => false,
+        }
```
Comment (mjibson) [inline src/adapter/src/session.rs:442]:
```suggestion
        matches!(self.transaction.inner(),
            Some(Transaction {
                pcx: _,
                ops: TransactionOps::Peeks(TimestampContext::TimelineTimestamp(_, _)),
                write_lock_guard: _,
                access: _,
                id: _,
            })
        )
```

I think this works?

File: src/adapter/src/coord/sequencer.rs [hunk 4]
```diff
@@ -2199,7 +2207,7 @@ impl<S: Append + 'static> Coordinator<S> {
         }
     }
 
-    fn plan_peek(
+    async fn plan_peek(
```
Comment (mjibson) [inline src/adapter/src/coord/sequencer.rs:2210]: Not thrilled that this an explain are now async. Are these getting fixed (either reverted, or else they now get a `tx` and handle responses themselves) in followups?

**Review by mjibson — APPROVED** [review]

File: src/adapter/src/coord/sequencer.rs (hunk 2 above)
Comment (jkosh44) [reply src/adapter/src/coord/sequencer.rs:2405]: Added some docs.

File: src/adapter/src/session.rs (hunk 3 above)
Comment (jkosh44) [reply src/adapter/src/session.rs]: Yep, updated.

File: src/adapter/src/coord/sequencer.rs (hunk 4 above)
Comment (jkosh44) [reply src/adapter/src/coord/sequencer.rs:2211]: Yeah I'm hoping to remove this in followups, I've added a TODO.

**Comment by jkosh44** [comment]
@petrosagg When this PR merges, the functionality for RTR will be fully implemented on the adapter side so storage can theoretically implement and test independently. Though we still need to make optimizations on the adapter side before it's production ready.

**Pull request merged**